hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -3684,32 +3625,96 @@ class Emit[C](val ctx: EmitContext, val cb: EmitClassBuilder[C]) {
       cb.memoize(cb.invokeCode[Boolean](sort, cb.this_, region, l, r))
   }
 
-  def emitLet[A](
-    emitI: (IR, EmitCodeBuilder, EmitEnv) => IEmitCode,
-    emitBody: (IR, EmitCodeBuilder, EmitEnv) => A,
+  def emitLetBindings(
+    emitI: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => IEmitCode,
+    emitVoid: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => Unit,
   )(
     let: Let,
     cb: EmitCodeBuilder,
     env: EmitEnv,
-  ): A = {
+    r: Value[Region],
+  ): EmitEnv = {
     val uses: mutable.Set[String] =
       ctx.usesAndDefs.uses.get(let) match {
         case Some(refs) => refs.map(_.t.name)
         case None => mutable.Set.empty
       }
 
-    emitBody(
-      let.body,
-      cb,
-      let.bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
-        if (!uses.contains(name)) newEnv
-        else {
-          val value = emitI(ir, cb, newEnv)
+    def emitChunk(cb: EmitCodeBuilder, bindings: Seq[(String, IR)], env: EmitEnv, r: Value[Region])
+      : EmitEnv =
+      bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
+        if (ir.typ == TVoid) {
+          emitVoid(ir, cb, newEnv, r)
+          newEnv
+        } else if (IsPure(ir) && !uses.contains(name)) {
+          newEnv
+        } else {
+          val value = emitI(ir, cb, newEnv, r)
           val memo = cb.memoizeMaybeStreamValue(value, s"let_$name")
           newEnv.bind(name, memo)
         }
-      },
-    )
+      }
+
+    @tailrec def go(
+      env: EmitEnv,
+      chunkStart: Int,
+      pos: Int,
+      chunkSize: Int,
+      groupIdx: Int,
+    ): EmitEnv = {
+
+      def emitChunkWrapped(): EmitEnv = {
+        val mb = cb.emb.genEmitMethod(
+          s"begin_group_$groupIdx",
+          FastSeq[ParamType](classInfo[Region]),
+          UnitInfo,
+        )
+        var newEnv = env
+        mb.voidWithBuilder { cb =>
+          newEnv =
+            emitChunk(cb, let.bindings.slice(chunkStart, pos), env, mb.getCodeParam[Region](1))
+        }
+        cb.invokeVoid(mb, cb.this_, r)
+        newEnv
+      }
+
+      if (pos == let.bindings.length) {
+        if (chunkSize > 0)
+          return emitChunkWrapped()
+        else
+          return env
+      }
+
+      val (curName, curIR) = let.bindings(pos)
+
+      // skip over unused streams
+      if (curIR.typ.isInstanceOf[TStream] && !uses.contains(curName)) {
```

Comment:
We've testing for some uses in more than two places, can these be folded into one?

Reply:
I think only by using an array builder to hold the current chunk, instead of just slicing into the `bindings` seq. `uses` is just a set, so testing for membership is fast.